http/src/main/scala/org/http4s/blaze/http/ClientSessionManagerImpl.scala

```diff
@@ -0,0 +1,167 @@
+package org.http4s.blaze.http
+
+import org.http4s.blaze.channel.nio2.ClientChannelFactory
+import org.http4s.blaze.http.ClientSessionManagerImpl._
+import org.http4s.blaze.http.HttpClientSession.{Closed, ReleaseableResponse, Status}
+import org.http4s.blaze.http.http1.client.Http1ClientStage
+import org.http4s.blaze.http.util.UrlTools.UrlComposition
+import org.http4s.blaze.pipeline.stages.SSLStage
+import org.http4s.blaze.pipeline.{Command, LeafBuilder}
+import org.http4s.blaze.util.Execution
+import org.log4s.getLogger
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Future, Promise}
+import scala.util.control.NonFatal
+import scala.util.{Failure, Success}
+
+private object ClientSessionManagerImpl {
+  case class ConnectionId(scheme: String, authority: String)
+
+  case class Http1SessionProxy(id: ConnectionId, parent: Http1ClientSession) extends Http1ClientSession {
+    override def dispatch(request: HttpRequest): Future[ReleaseableResponse] = parent.dispatch(request)
+    override def status: Status = parent.status
+    override def close(within: Duration): Future[Unit] = parent.close(within)
+  }
+
+  def apply(config: HttpClientConfig): ClientSessionManagerImpl = {
+    val sessionCache = new java.util.HashMap[ConnectionId, java.util.Collection[HttpClientSession]]
+    new ClientSessionManagerImpl(sessionCache, config)
+  }
+}
+
+private final class ClientSessionManagerImpl(sessionCache: java.util.Map[ConnectionId, java.util.Collection[HttpClientSession]], config: HttpClientConfig) extends ClientSessionManager {
+
+  private[this] def getId(composition: UrlComposition): ConnectionId =
+    ConnectionId(composition.scheme, composition.authority)
+
+  private[this] implicit def ec = Execution.trampoline
+
+  private[this] val logger = getLogger
+  private[this] val socketFactory = new ClientChannelFactory(group = config.channelGroup)
+
+  override def acquireSession(request: HttpRequest): Future[HttpClientSession] = {
+    logger.debug(s"Acquiring session for request $request")
+     UrlComposition(request.url) match {
+      case Success(urlComposition) =>
+        val id = getId(urlComposition)
+        val session = findExistingSession(id)
+
+        if (session == null) createNewSession(urlComposition, id)
+        else {
+          logger.debug(s"Found hot session for id $id: $session")
+          Future.successful(session)
+        }
+
+      case Failure(t) => Future.failed(t)
+    }
+  }
+
+  // WARNING: can emit `null`. For internal use only
+  private[this] def findExistingSession(id: ConnectionId): HttpClientSession = sessionCache.synchronized {
```

Comment:
What are we getting out of a synchronized map instead of a concurrent map?  Is it the side effects that might be performed in the collection under the key?

Reply:
It seemed easier to reason about than having a concurrent map full of concurrent lists.